### PR TITLE
Power Panic and LCD menu

### PR DESCRIPTION
**Pull request by 3d-gussner**
Improved Power Panic
- Added Debug for printer states
- SD power panic
  - Recovers from multiple power panic and even resets (as long it is in recovering)
- Host power panic
- LCD menu
  - During recovering all menus are hidden except
    - Stop print (to be able to exit)
    - Statistics
    - Fail Stats
    - Support   
  - Reprint hidden
    - Host reprint only shown when host sends `M79` to ensure that the host is ready to reprint
    - SD Reprint only shown when previous print was finished successfully.
      - ToDo for next release: Don't show Reprint when SD card has been changed
  - Resume hidden
    - When resuming
    - With host prints until host sends `M79`
  - `Autoload filament` menu only shown with active FSensor and NO filament detected
  - `Load filament` menu only shown with active FSensor and NO filament detected OR FSensor disabled
  - `Unload filament` menu only shown with active FSensor and filament detected OR FSensor disabled
 
 
This Firmware doesn't recover correctly with Prusa Link 0.8.0-alpha4 at this moment.
I was able to recover host prints with a modified OctoPrint + two Plugins and additional code (which isn't in this PR as this code is a workaround of existing OctoPrint+Plugins)

Update 27-03-2024

Added `EEPROM_UVLO_Z_LIFTED` to prevent multiple Z lifts in case of a power panic while the print is paused.
A host print recovery it doesn't wait for hotend temp to be stable as it will go into pause state anyway and the hotend will be disabled while paused anyway. 

Tested with MK404

- [X] SD print + power panic
  - For some unknown reason it doesn't recover correctly SD print + power panic + reset while recovering, while a double power panic does recover correctly even with a reset.   
- [X] SD print + power panic and while lifting Z another power panic
- [X] SD print + power panic and while lifting Z another power panic + reset while recovering
- [X] SD print + pause + power panic
- [X] SD print + pause + power panic and while lifting Z another power panic
- [X] SD print + pause + power panic and while lifting Z another power panic + reset while recovering
- [X] Host print same as above for SD

Difference between SD and Host print power panic:
- SD recovers automatically when bed temp isn't lower than 5°C to target bed temperature
- Host recovers the print settings and goes to pause state.
  - Then it waits for the host sending "M79" before showing Resume to ensure that the host is "there" to receive any messages from the printer.  
   
The print can't be recovered if a reset happens at the last step of the recovery as at this moment the power panic recovery is "completed" and the printer is back in normal mode.

### CODE REVIEW

File: Firmware/ultralcd.cpp [hunk 1]
```diff
@@ -5228,6 +5228,10 @@ static void lcd_main_menu()
         MENU_ITEM_FUNCTION_P(PSTR("tst - Restore"), lcd_menu_test_restore);
 #endif //RESUME_DEBUG
 
+#ifdef DEBUG_PRINTER_STATES
+    debug_printer_states();
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp:5232]: I have a bad feeling about this. `lcd_main_menu()` is called very frequently and dumping this much stuff on the serial stream will likely do more harm than good. Does the printer actually keep up with this much serial prints?

Comment (gudnimg) [reply Firmware/ultralcd.cpp:5232]: Another point is, what happens when the printer is not rendering the main menu? Then these debug prints are ignored.

File: Firmware/Marlin_main.cpp [hunk 2]
```diff
@@ -4117,62 +4173,52 @@ void process_commands()
...
-        else if (code_seen_P(PSTR("FAN"))) { // PRUSA FAN
+        } else if (code_seen_P(PSTR("FAN"))) { // PRUSA FAN
             printf_P(_N("E0:%d RPM\nPRN0:%d RPM\n"), 60*fan_speed[0], 60*fan_speed[1]);
-        }
-        else if (code_seen_P(PSTR("uvlo"))) // PRUSA uvlo
-        {
-            if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_SD)
-            {
+        } else if (code_seen_P(PSTR("uvlo"))) { // PRUSA uvlo
+            if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_SD) {
```
Comment (gudnimg) [inline Firmware/Marlin_main.cpp:4180]: Can `saved_printing_type` be used here?

`if (saved_printing_type == PowerPanic::PRINT_TYPE_SD)`

File: Firmware/Marlin_main.cpp [hunk 3]
```diff
@@ -4117,62 +4173,52 @@ void process_commands()
...
                 // M24 - Start SD print
                 enquecommand_P(MSG_M24);
 
                 // Print is recovered, clear the recovery flag
                 eeprom_update_byte_notify((uint8_t*)EEPROM_UVLO, PowerPanic::NO_PENDING_RECOVERY);
-            }
-            else if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_HOST)
-            {
+            } else if (eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE) == PowerPanic::PRINT_TYPE_HOST) {
```
Comment (gudnimg) [inline Firmware/Marlin_main.cpp:4187]: Can `saved_printing_type` be used here?

`if (saved_printing_type == PowerPanic::PRINT_TYPE_HOST)`

File: Firmware/Marlin_main.cpp [hunk 4]
```diff
@@ -11024,7 +11069,7 @@ void restore_print_from_ram_and_continue(float e_move)
 	set_destination_to_current();
 
     restore_print_file_state();
-
+    eeprom_update_byte((uint8_t*)EEPROM_UVLO, PowerPanic::NO_PENDING_RECOVERY);
 	lcd_setstatuspgm(MSG_WELCOME);
     saved_printing_type = PowerPanic::PRINT_TYPE_NONE;
 	saved_printing = false;
```
Comment (gudnimg) [inline Firmware/Marlin_main.cpp]: Optimisation idea: Can we use `cancel_saved_printing()` here?

So change:

```c
    eeprom_update_byte((uint8_t*)EEPROM_UVLO, PowerPanic::NO_PENDING_RECOVERY);
	lcd_setstatuspgm(MSG_WELCOME);
    saved_printing_type = PowerPanic::PRINT_TYPE_NONE;
	saved_printing = false;
```

to this:

```c
    cancel_saved_printing();
	lcd_setstatuspgm(MSG_WELCOME);
```

File: Firmware/Marlin_main.cpp [hunk 5]
```diff
@@ -1581,32 +1636,33 @@ void setup()
...
       manage_heater(); // Update temperatures 
-#ifdef DEBUG_UVLO_AUTOMATIC_RECOVER 
-		printf_P(_N("Power panic detected!\nCurrent bed temp:%d\nSaved bed temp:%d\n"), (int)degBed(), eeprom_read_byte((uint8_t*)EEPROM_UVLO_TARGET_BED));
-#endif
-     uvlo_auto_recovery_ready = (degBed() > ( (float)eeprom_read_byte((uint8_t*)EEPROM_UVLO_TARGET_BED) - AUTOMATIC_UVLO_BED_TEMP_OFFSET));
-     if (uvlo_auto_recovery_ready){
-          #ifdef DEBUG_UVLO_AUTOMATIC_RECOVER 
+      //Restore printing type
+      saved_printing_type = eeprom_read_byte((uint8_t*)EEPROM_UVLO_PRINT_TYPE);
+#ifdef DEBUG_UVLO_AUTOMATIC_RECOVER
```
Comment (gudnimg) [inline Firmware/Marlin_main.cpp:1644]: Personally I really don't like this `DEBUG_UVLO_AUTOMATIC_RECOVER` code. It makes the code very unreadable. We can probably move all the serial print messages into `recover_print()` function.

Just an observation/comment. Not a blocker.

File: Firmware/power_panic.cpp [hunk 6]
```diff
@@ -315,11 +315,17 @@ void recover_print(uint8_t automatic) {
...
     bool mbl_was_active = recover_machine_state_after_power_panic();
 
-    // Lift the print head 25mm, first to avoid collisions with oozed material with the print,
-    // and second also so one may remove the excess priming material.
+    // Undo PP Z Lift by setting current Z pos to + Z_PAUSE_LIFT
+    // After a reboot the printer doesn't know the Z height and we have to set its previous value
+    if(eeprom_read_byte((uint8_t*)EEPROM_UVLO) == PowerPanic::PENDING_RECOVERY_RETRY) {
+        current_position[Z_AXIS] += Z_PAUSE_LIFT;
+    }
+
```
Comment (gudnimg) [inline Firmware/power_panic.cpp]: I'm not sure about this change. `current_position` is likely not modified unless the g-code was executed.

`enquecommandf_P(PSTR("G1 Z%.3f F800"), current_position[Z_AXIS] + Z_PAUSE_LIFT);`

Doesn't the if statement below make sure this lift isn't executed again?:

`if(eeprom_read_byte((uint8_t*)EEPROM_UVLO) == PowerPanic::PENDING_RECOVERY)`

There shouldn't be another lift when the flag is set to `PowerPanic::PENDING_RECOVERY_RETRY`. Am I missing something?

**Review by gudnimg — CHANGES_REQUESTED** [review]
See my comments

File: Firmware/ultralcd.cpp [hunk 7]
```diff
@@ -5297,81 +5314,96 @@ static void lcd_main_menu()
 #endif
 
     // only allow starting SD print if hardware errors (temperature or fan) are cleared
-    if(!get_temp_error()
-            && !printer_active()
+    if (!printer_recovering()
+        && !printer_active()
+        && !get_temp_error()
+        && !printer_active()
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: Redundant call to `!printer_active()` here

File: Firmware/ultralcd.cpp [hunk 8]
```diff
@@ -5297,81 +5314,96 @@ static void lcd_main_menu()
...
-    }
-
-    if(!printer_active() && !farm_mode) {
-        const int8_t sheet = eeprom_read_byte(&(EEPROM_Sheets_base->active_sheet));
-        const int8_t nextSheet = eeprom_next_initialized_sheet(sheet);
-        if ((nextSheet >= 0) && (sheet != nextSheet)) { // show menu only if we have 2 or more sheets initialized
-            MENU_ITEM_FUNCTION_E(EEPROM_Sheets_base->s[sheet], eeprom_switch_to_next_sheet);
-        }
+        if(!printer_active()
+            && !farm_mode) {
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: Why is this split into two lines? Can easily fit one line and is more readable IMO

File: Firmware/ultralcd.cpp [hunk 9]
```diff
@@ -5297,81 +5314,96 @@ static void lcd_main_menu()
...
+            const int8_t sheet = eeprom_read_byte(&(EEPROM_Sheets_base->active_sheet));
+            const int8_t nextSheet = eeprom_next_initialized_sheet(sheet);
+            if ((nextSheet >= 0)
+                && (sheet != nextSheet)) { // show menu only if we have 2 or more sheets initialized
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: Similar here. Why split this into two lines. I feel it is very unusual.
The comment can be either above the if statement or inside it.

File: Firmware/ultralcd.cpp (hunk 1 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:5232]: You are correct, this menu will dump lot of information over serial and should only be used with the MK404 or with a subset of the messages.
I added this to be able to troubleshoot some LCD menu's.

File: Firmware/Marlin_main.cpp (hunk 2 above)
Comment (3d-gussner) [reply Firmware/Marlin_main.cpp:4180]: When you assume that `PRUSA uvlo` is only used by the "normal" Power Panic recovery functions then the `saved_printing_type` is already read before the `PRUSA uvlo`  is called.
But as this is a gcode that can be executed any time the `saved_printing_type` may not be correct.
I have no idea why this gcode even exists.

File: Firmware/Marlin_main.cpp (hunk 4 above)
Comment (3d-gussner) [reply Firmware/Marlin_main.cpp]: No as `cancle_saved_printing()` sets `saved_start_position[0] = SAVED_START_POSITION_UNSET;` and the function 
`void restore_print_from_ram_and_continue(float e_move)` only resets some states BUT continues to print.

File: Firmware/Marlin_main.cpp (hunk 5 above)
Comment (3d-gussner) [reply Firmware/Marlin_main.cpp:1644]: It is 6 years old and as DEBUG code has been placed close to the operation you are trying to debug.
In VScode and an existing `Configuration_prusa.h` it is "grayed out" and so not that disturbing to read at least for me.

File: Firmware/power_panic.cpp (hunk 6 above)
Comment (3d-gussner) [reply Firmware/power_panic.cpp]: Fixed with EEPROM_UVLO_Z_LIFTED

File: Firmware/ultralcd.cpp [hunk 10]
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
...
+             MENU_ITEM_GCODE_P(_i("Change SD card"), PSTR("M21"));  // SD-card changed by user ////MSG_CNG_SDCARD c=18
 #endif //SDCARDDETECT
-            }
-        } else {
-            bMain=true;                                   // flag (i.e. 'fake parameter') for 'lcd_sdcard_menu()' function
-            MENU_ITEM_SUBMENU_P(_i("No SD card"), lcd_sdcard_menu); ////MSG_NO_CARD c=18
+                }
+            } else {
+                bMain=true;                                   // flag (i.e. 'fake parameter') for 'lcd_sdcard_menu()' function
+                MENU_ITEM_BACK_P(_i("No SD card")); ////MSG_NO_CARD c=18
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp:5321]: This changes the UI, is this change needed?

File: Firmware/ultralcd.cpp [hunk 11]
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
 #endif
 
     // only allow starting SD print if hardware errors (temperature or fan) are cleared
-    if(!get_temp_error()
-            && !printer_active()
+    if (!printer_recovering() && !printer_active() && !get_temp_error()
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp:5302]: Potential optimisation: there are a few if statements inside this top-level if statements which check for `!printer_active()`. That shouldn't be needed anymore since this top-level if statement checks it.

File: Firmware/ultralcd.cpp [hunk 12]
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
...
 #endif //SDSUPPORT
-    }
-
-    if(!printer_active() && !farm_mode) {
-        const int8_t sheet = eeprom_read_byte(&(EEPROM_Sheets_base->active_sheet));
-        const int8_t nextSheet = eeprom_next_initialized_sheet(sheet);
-        if ((nextSheet >= 0) && (sheet != nextSheet)) { // show menu only if we have 2 or more sheets initialized
-            MENU_ITEM_FUNCTION_E(EEPROM_Sheets_base->s[sheet], eeprom_switch_to_next_sheet);
-        }
+        if(!printer_active() && !farm_mode) {
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: `!printer_active()` should always be true here since top-level if statement checks for it

File: Firmware/ultralcd.cpp [hunk 13]
```diff
@@ -5297,81 +5299,87 @@ static void lcd_main_menu()
...
-    }
+        }
 
-    if ( ! ( printer_active() || (eFilamentAction != FilamentAction::None) || Stopped ) ) {
-        if (MMU2::mmu2.Enabled()) {
-            if(!MMU2::mmu2.FindaDetectsFilament() && !fsensor.getFilamentPresent()) {
-                // The MMU 'Load filament' state machine will reject the command if any 
-                // filament sensor is reporting a detected filament
-                MENU_ITEM_SUBMENU_P(_T(MSG_PRELOAD_TO_MMU), mmu_preload_filament_menu);
+        if (!( printer_active() || (eFilamentAction != FilamentAction::None) || Stopped )) {
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp]: `!printer_active()` should always be true here since top-level if statement checks for it

File: Firmware/ultralcd.cpp [hunk 14]
```diff
@@ -5249,45 +5253,43 @@ static void lcd_main_menu()
...
+        && fan_check_error != EFCE_REPORTED
 #endif //FANCHECK
-           ) {
-            if (saved_printing) {
+        && (saved_printing_type != PowerPanic::PRINT_TYPE_NONE || saved_printing)
+        && custom_message_type != CustomMsg::Resuming) {
+        if (saved_printing_type == PowerPanic::PRINT_TYPE_SD) {
                 MENU_ITEM_SUBMENU_P(_T(MSG_RESUME_PRINT), lcd_resume_print);
-            } else {
+        } else if ((saved_printing_type == PowerPanic::PRINT_TYPE_HOST) && (M79_timer_get_status())) {
```
Comment (gudnimg) [inline Firmware/ultralcd.cpp:5287]: hmm, isn't this a regression for users which **don't** use the M79 timer? If a user doesn't use M79, this change means they can no longer resume the print from the LCD. But still they can pause the print from LCD.

See `lcd_pause_usb_print` which doesn't have this restriction.

**Review by gudnimg — CHANGES_REQUESTED** [review]
See comments

File: Firmware/ultralcd.cpp (hunk 13 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp]: Right, thanks

File: Firmware/ultralcd.cpp (hunk 12 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp]: Right, thanks

File: Firmware/ultralcd.cpp (hunk 11 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:5302]: Right, thanks

**Review by gudnimg — APPROVED** [review]

File: Firmware/ultralcd.cpp (hunk 14 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:5287]: It has been pointed out that the menu shouldn't be shown as long the host isn't connected.
Testing selected Resume too early while the RPi Zero was still booting up after a power panic. Nothing happened as the `//action:resume` never reached the host.
Using the `M79` ensures that the host is connected and ready to receive instructions like `//action:resume` and prevent users to select the menu too early.
